bids/layout/models.py

```diff
@@ -32,35 +28,60 @@ def __init_subclass__(cls, extension, **kwargs):
             BIDSFile._ext_registry[ext] = cls
 
     @classmethod
-    def from_path(cls, path):
-        path = Path(path)
+    def from_filename(cls, filename):
+        path = Path(filename)
         for ext, subclass in cls._ext_registry.items():
             if path.name.endswith(ext):
                 cls = subclass
                 break
-        return cls(path)
+        return cls(filename=path)
 
-    def __init__(self, filename, root=None):
-        self.path = str(filename)
-        self.filename = self._path.name
-        self.dirname = str(self._path.parent)
-        self.is_dir = not self.filename
-        self._root = root
+    @classmethod
+    def from_artifact(cls, artifact):
+        """ Load from ANCPBids Artifact """
+        return cls(artifact=artifact)
+
+    def __init__(self, artifact=None, filename=None):
```

Comment:
I would suggest the following approach:

```Python
class BIDSFile:
    # Take a single argument, and decide if it's an artifact or not
    def __init__(self, file_ref: str | os.PathLike | Artifact):
        self._path = None
        self._artifact = None
        if isinstance(file_ref, (str, os.PathLike)):
            self._path = Path(file_ref)
        elif isinstance(file_ref, Artifact):
            self._artifact = file_ref

    # For essential functionality, try to use artifact (common case)
    # Fall back to Path if possible (otherwise raise NotImplementedError)
    @property
    def path(self):
        try:
            return self._artifact.name
        except AttributeError:
            return str(self._path)

    # As much as possible call things like `self.path` to keep the branches localized to a few places
```

I would not expose an approach to create an `Artifact`-backed `BIDSFile` to the user because then we have to support it. If needed for unit tests, we can create a function to save ourselves time in `bids.testing` or some such:

```Python
def make_BIDSFile(path):
    return BIDSFile(Artifact(**parse_bids_name(filename)))
```

Reply:
That works, although functionality would be fairly limited if constructed from a path. For example, we would not have access to `.get_entities ` or `.get_metadata`. Can't think of a use case for a `BIDSFile` without those properties

Reply:
I push an implementation of your suggestion--- but I would be fine dropping filepath construction of `BIDSFile`. It unnecessarily complicates the object.

Reply:
Ready for another review @effigies

I think this can be merged if no problems